applications/serial_lte_modem/src/slm_at_socket.c

```diff
@@ -0,0 +1,1579 @@
+				break;
+			}
+		} else {
+			offset += ret;
+		}
+	}
+
+	return offset;
+}
+
+static int do_recvfrom(int timeout)
+{
+	int ret;
+	struct sockaddr remote;
+	socklen_t addrlen = sizeof(struct sockaddr);
+	char rx_data[SLM_MAX_PAYLOAD];
+	int length;
+
+	if (sock.family == AF_INET) {
+		length = UDP_MAX_PAYLOAD_IPV4;
+	} else {
+		length = UDP_MAX_PAYLOAD_IPV6;
+	}
+	ret = socket_poll(sock.fd, POLLIN, timeout);
+	if (ret) {
+		return ret;
+	}
+	ret = recvfrom(sock.fd, (void *)rx_data, length, 0, &remote, &addrlen);
+	if (ret < 0) {
+		LOG_ERR("recvfrom() error: %d", -errno);
+		return -errno;
+	}
+	/**
+	 * Datagram sockets in various domains permit zero-length
+	 * datagrams. When such a datagram is received, the return
+	 * value is 0. Treat as normal case
+	 */
+	if (ret == 0) {
+		LOG_WRN("recvfrom() return 0");
+	} else {
+		char peer_addr[NET_IPV6_ADDR_LEN] = {0};
+
+		if (remote.sa_family == AF_INET) {
+			(void)inet_ntop(AF_INET, &((struct sockaddr_in *)&remote)->sin_addr,
+			    peer_addr, sizeof(peer_addr));
+		} else if (remote.sa_family == AF_INET6) {
+			(void)inet_ntop(AF_INET6, &((struct sockaddr_in6 *)&remote)->sin6_addr,
+			    peer_addr, sizeof(peer_addr));
+		}
+		rsp_send(rx_data, ret);
+		sprintf(rsp_buf, "\r\n#XRECVFROM: %d,\"%s\"\r\n", ret, peer_addr);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	}
+
+	return 0;
+}
+
+static int socket_poll(int sock_fd, int event, int timeout)
+{
+	int ret;
+	struct pollfd fd = {
+		.fd = sock_fd,
+		.events = event
+	};
+
+	if (timeout <= 0) {
+		return 0;
+	}
+
+	ret = poll(&fd, 1, MSEC_PER_SEC * timeout);
+	if (ret < 0) {
+		LOG_WRN("poll() error: %d", -errno);
+		return -errno;
+	} else if (ret == 0) {
+		LOG_WRN("poll() timeout");
+		return -EAGAIN;
+	}
+
+	LOG_DBG("poll() events 0x%08x", fd.revents);
+	if ((fd.revents & event) != event) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+static int socket_datamode_callback(uint8_t op, const uint8_t *data, int len)
+{
+	int ret = 0;
+
+	if (op == DATAMODE_SEND) {
+		if (strlen(udp_url) > 0) {
+			ret = do_sendto_datamode(data, len);
+		} else {
+			ret = do_send_datamode(data, len);
+		}
+		LOG_INF("datamode send: %d", ret);
+	} else if (op == DATAMODE_EXIT) {
+		LOG_DBG("datamode exit");
+		memset(udp_url, 0x00, sizeof(udp_url));
+	}
+
+	return ret;
+}
+
+/**@brief handle AT#XSOCKET commands
+ *  AT#XSOCKET=<op>[,<type>,<role>]
+ *  AT#XSOCKET?
+ *  AT#XSOCKET=?
+ */
+int handle_at_socket(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t op;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_unsigned_short_get(&at_param_list, 1, &op);
+		if (err) {
+			return err;
+		}
+		if (op == AT_SOCKET_OPEN || op == AT_SOCKET_OPEN6) {
+			err = at_params_unsigned_short_get(&at_param_list, 2, &sock.type);
+			if (err) {
+				return err;
+			}
+			err = at_params_unsigned_short_get(&at_param_list, 3, &sock.role);
+			if (err) {
+				return err;
+			}
+			if (sock.fd > 0) {
```

Comment:
Actually, 0 is still a valid file descriptor, so it should be >= 0. Or simply check as in other places with `if (sock.fd != INVALID_SOCKET)`

Reply:
Yes, thanks this is an error